File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h10-w/low_level/low_state.py
#!/usr/bin/env python3

import humanoid_sdk_py
import humanoid_sdk_py.h10w as h10w
import time


def main():
    print("=== Low State Reading Test ===")

    low_level = h10w.H10wLowLevel()

    DURATION_MS = 5000
    READ_INTERVAL_MS = 100
    read_count = DURATION_MS // READ_INTERVAL_MS

    for i in range(read_count):
        start_time = time.time()

        ret, low_state = low_level.readLowState()
        if ret != 0:
            print("Failed to read low state!")
            return -1

        print(f"=== Low State Reading {i + 1} ===")
        print(f"Version: {low_state.version[0]}.{low_state.version[1]}")
        print(f"Tick: {low_state.tick}")
        print(f"Data valid: {low_state.data_valid}")

        if low_state.data_valid > 0:
            j0 = low_state.joint_motor_state[0]
            print(
                "Joint 0 - Mode: {}, Op mode: {}, Enable: {}, "
                "Position: {}, Velocity: {}, Acceleration: {}, "
                "Torque: {}, Current: {}, Voltage: {}, "
                "Temp: [{}, {}], Kp: {}, Kd: {}, Error: {}, Warning: {}".format(
                    j0.mode,
                    j0.op_mode,
                    j0.enable_state,
                    j0.q,
                    j0.dq,
                    j0.ddq,
                    j0.tau_est,
                    j0.cur,
                    j0.vol,
                    j0.temp[0],
                    j0.temp[1],
                    j0.kp,
                    j0.kd,
                    j0.error_code,
                    j0.warning_code,
                )
            )

            w0 = low_state.wheel_motor_state[0]
            print(
                "Wheel 0 - Mode: {}, Op mode: {}, Enable: {}, "
                "Velocity: {}, Torque: {}, Current: {}, Voltage: {}, "
                "Temp: [{}, {}], Error: {}, Warning: {}".format(
                    w0.mode,
                    w0.op_mode,
                    w0.enable_state,
                    w0.v,
                    w0.tau_est,
                    w0.cur,
                    w0.vol,
                    w0.temp[0],
                    w0.temp[1],
                    w0.error_code,
                    w0.warning_code,
                )
            )

            imu = low_state.imu_state
            print(
                "IMU - Accel: [{}, {}, {}], Gyro: [{}, {}, {}], "
                "Temp: {}, Error: {}, Warning: {}".format(
                    imu.accelerometer[0],
                    imu.accelerometer[1],
                    imu.accelerometer[2],
                    imu.gyroscope[0],
                    imu.gyroscope[1],
                    imu.gyroscope[2],
                    imu.temp,
                    imu.error_code,
                    imu.warning_code,
                )
            )

        pcu = low_state.pcu_state
        print(
            "Power state: {}, Emergency stop: {}, Soft emergency stop: {}, "
            "Bumper[0]: {}, PCU Error: {}, PCU Warning: {}".format(
                pcu.power_state,
                pcu.emergency_stop_state,
                pcu.soft_emergency_stop_state,
                pcu.bumper_switch_state[0],
                pcu.error_code,
                pcu.warning_code,
            )
        )
        en = low_state.enable_state
        print(
            "Upper enable state: {}, Base enable state: {}".format(
                en.upper_enable_state,
                en.base_enable_state,
            )
        )
        print(f"Time seconds: {low_state.time_seconds}")
        print(f"Error code: {low_state.error_code}")
        print(f"Warning code: {low_state.warning_code}")
        print()

        elapsed_ms = (time.time() - start_time) * 1000.0
        sleep_ms = READ_INTERVAL_MS - elapsed_ms
        if sleep_ms > 0:
            time.sleep(sleep_ms / 1000.0)

    print("Low state reading test completed successfully")
    return 0


if __name__ == "__main__":
    exit(main())

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h10-w/low_level/low_state.cpp
// low_state.cpp

#include <chrono>
#include <iostream>
#include <thread>

#include "common/common_def.h"
#include "common/h10w_def.h"
#include "h10-w/h10w_low_level.h"

using namespace humanoid;
using namespace humanoid::h10w;

int main() {
  std::cout << "=== Low State Reading Test ===" << std::endl;

  H10wLowLevel low_level;
  LowState low_state;

  // Read low state for 5 seconds
  constexpr int DURATION_MS = 5000;
  constexpr int READ_INTERVAL_MS = 100;
  constexpr int read_count = DURATION_MS / READ_INTERVAL_MS;

  for (int i = 0; i < read_count; ++i) {
    auto start = std::chrono::high_resolution_clock::now();

    int ret = low_level.readLowState(low_state);
    if (ret != 0) {
      std::cout << "Failed to read low state!" << std::endl;
      return -1;
    }

    // Print some state information
    std::cout << "=== Low State Reading " << i + 1 << " ===" << std::endl;
    std::cout << "Version: " << low_state.version[0] << "."
              << low_state.version[1] << std::endl;
    std::cout << "Tick: " << low_state.tick << std::endl;
    std::cout << "Data valid: " << static_cast<int>(low_state.data_valid)
              << std::endl;

    // Print first joint state as example
    if (low_state.data_valid > 0) {
      const auto& joint0 = low_state.joint_motor_state[0];
      std::cout << "Joint 0 - Mode: " << static_cast<int>(joint0.mode)
                << ", Op mode: " << static_cast<int>(joint0.op_mode)
                << ", Enable: " << static_cast<int>(joint0.enable_state)
                << ", Position: " << joint0.q << ", Velocity: " << joint0.dq
                << ", Acceleration: " << joint0.ddq
                << ", Torque: " << joint0.tau_est << ", Current: " << joint0.cur
                << ", Voltage: " << joint0.vol << ", Temp: [" << joint0.temp[0]
                << ", " << joint0.temp[1] << "], Kp: " << joint0.kp
                << ", Kd: " << joint0.kd << ", Error: " << joint0.error_code
                << ", Warning: " << joint0.warning_code << std::endl;
      const auto& wheel0 = low_state.wheel_motor_state[0];
      std::cout << "Wheel 0 - Mode: " << static_cast<int>(wheel0.mode)
                << ", Op mode: " << static_cast<int>(wheel0.op_mode)
                << ", Enable: " << static_cast<int>(wheel0.enable_state)
                << ", Velocity: " << wheel0.v << ", Torque: " << wheel0.tau_est
                << ", Current: " << wheel0.cur << ", Voltage: " << wheel0.vol
                << ", Temp: [" << wheel0.temp[0] << ", " << wheel0.temp[1]
                << "], Error: " << wheel0.error_code
                << ", Warning: " << wheel0.warning_code << std::endl;

      const auto& imu = low_state.imu_state;
      std::cout << "IMU - Accel: [" << imu.accelerometer[0] << ", "
                << imu.accelerometer[1] << ", " << imu.accelerometer[2] << "]"
                << ", Gyro: [" << imu.gyroscope[0] << ", " << imu.gyroscope[1]
                << ", " << imu.gyroscope[2] << "], Temp: " << imu.temp
                << ", Error: " << imu.error_code
                << ", Warning: " << imu.warning_code << std::endl;
    }

    std::cout << "Power state: "
              << static_cast<int>(low_state.pcu_state.power_state)
              << ", Emergency stop: "
              << static_cast<int>(low_state.pcu_state.emergency_stop_state)
              << ", Soft emergency stop: "
              << static_cast<int>(low_state.pcu_state.soft_emergency_stop_state)
              << ", Bumper[0]: "
              << static_cast<int>(low_state.pcu_state.bumper_switch_state[0])
              << ", PCU Error: " << low_state.pcu_state.error_code
              << ", PCU Warning: " << low_state.pcu_state.warning_code
              << std::endl;
    std::cout << "Upper enable state: "
              << static_cast<int>(low_state.enable_state.upper_enable_state)
              << ", Base enable state: "
              << static_cast<int>(low_state.enable_state.base_enable_state)
              << std::endl;
    std::cout << "Time seconds: " << low_state.time_seconds << std::endl;
    std::cout << "Error code: " << low_state.error_code << std::endl;
    std::cout << "Warning code: " << low_state.warning_code << std::endl;
    std::cout << std::endl;

    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed =
        std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::this_thread::sleep_for(std::chrono::milliseconds(READ_INTERVAL_MS) -
                                elapsed);
  }

  std::cout << "Low state reading test completed successfully" << std::endl;
  return 0;
}
